openprescribing/matrixstore/management/commands/matrixstore_init.py

```diff
@@ -0,0 +1,131 @@
+
+from gcutils.bigquery import Client
+from matrixstore.common import get_temp_filename
+from matrixstore.dates import generate_dates
+
+
+SCHEMA_SQL = """
+    CREATE TABLE presentation (
+        bnf_code TEXT,
+        is_generic BOOLEAN,
+        adq_per_quantity FLOAT,
+        name TEXT,
+        -- The below columns will contain the actual prescribing data as
+        -- serialized matrices of shape (number of practices, number of months)
+        items BLOB,
+        quantity BLOB,
+        actual_cost BLOB,
+        net_cost BLOB,
+
+        PRIMARY KEY (bnf_code)
+    );
+
+    CREATE TABLE practice_statistic (
+        name TEXT,
+        -- The "value" column will contain the actual statistics as serialized
+        -- matrices of shape (number of practices, number of months)
+        value BLOB,
+
+        PRIMARY KEY (name)
+    );
+
+    -- Maps each practice code to its corresponding row offset in the data matrix
+    CREATE TABLE practice (
+        offset INTEGER,
+        code TEXT UNIQUE,
+
+        PRIMARY KEY (offset)
+    );
+
+    -- Maps each date to its corresponding column offset in the data matrix
+    CREATE TABLE date (
+        offset INTEGER,
+        date TEXT UNIQUE,
+
+        PRIMARY KEY (offset)
+    );
+"""
+
+
+class Command(BaseCommand):
+    help = __doc__
+
+    def add_arguments(self, parser):
+        parser.add_argument('end_date', help='YYYY-MM format')
+        parser.add_argument('sqlite_path')
+
+    def handle(self, end_date, sqlite_path, **kwargs):
+        matrixstore_init(end_date, sqlite_path)
+
+
+def matrixstore_init(end_date, sqlite_path):
+    if os.path.exists(sqlite_path):
+        raise RuntimeError('File already exists at: ' + sqlite_path)
+    sqlite_path = os.path.abspath(sqlite_path)
+    temp_filename = get_temp_filename(sqlite_path)
+    sqlite_conn = sqlite3.connect(temp_filename)
+    bq_conn = Client('hscic')
+    sqlite_conn.executescript(SCHEMA_SQL)
+    dates = generate_dates(end_date)
+    import_dates(sqlite_conn, dates)
+    import_practices(bq_conn, sqlite_conn, dates)
+    import_presentations(bq_conn, sqlite_conn)
+    sqlite_conn.commit()
+    sqlite_conn.close()
+    os.rename(temp_filename, sqlite_path)
+
+
+def import_dates(sqlite_conn, dates):
+    sqlite_conn.executemany(
+        'INSERT INTO date (offset, date) VALUES (?, ?)',
+        enumerate(dates)
+    )
+
+
+def import_practices(bq_conn, sqlite_conn, dates):
+    # We only create practice entries for practices which have prescribed in
+    # the current date range.  The stored matrices are sparse, so having
+    # "empty" practice rows doesn't make much difference there. But when we
+    # start summing and processing these we use dense matrices and so it's
+    # better to cut down the number of rows to a minimum.
+    sql = (
+        """
+        SELECT DISTINCT practice FROM {hscic}.prescribing
+          WHERE month >= '%s' AND month <= '%s'
+          ORDER BY practice
```

Comment:
What about filtering by `setting=4`?

Reply:
I think we decided that we might still be interested in prescribing outside of standard GP practices. For instance, we might try to build local authority dashboards.